### ISSUE TITLE
Cannot import RealmSwift via SPM

### ISSUE THREAD

**Issue by m-petes** [cgrindel/rules_swift_package_manager#808]
After adding this to Package.swift
```
        .package(url: "https://github.com/realm/realm-swift.git", from: "10.17.0"), 
```

We see the following error
```
INFO: Repository rules_swift_package_manager~0.14.0~swift_deps~swiftpkg_realm_core instantiated at:
  callstack not available
Repository rule swift_package defined at:
  /private/var/tmp/_bazel_krypt/de9844abaa047ea4e0868fa95089f211/external/rules_swift_package_manager~0.14.0/swiftpkg/internal/swift_package.bzl:165:32: in <toplevel>
ERROR: An error occurred during the fetch of repository 'rules_swift_package_manager~0.14.0~swift_deps~swiftpkg_realm_core':
   Traceback (most recent call last):
	File "/private/var/tmp/_bazel_krypt/de9844abaa047ea4e0868fa95089f211/external/rules_swift_package_manager~0.14.0/swiftpkg/internal/swift_package.bzl", line 65, column 28, in _swift_package_impl
		pkg_ctx = pkg_ctxs.read(repository_ctx, directory, env)
	File "/private/var/tmp/_bazel_krypt/de9844abaa047ea4e0868fa95089f211/external/rules_swift_package_manager~0.14.0/swiftpkg/internal/pkg_ctxs.bzl", line 13, column 28, in _read
		pkg_info = pkginfos.get(
	File "/private/var/tmp/_bazel_krypt/de9844abaa047ea4e0868fa95089f211/external/rules_swift_package_manager~0.14.0/swiftpkg/internal/pkginfos.bzl", line 87, column 39, in _get
		desc_manifest = _get_desc_manifest(
	File "/private/var/tmp/_bazel_krypt/de9844abaa047ea4e0868fa95089f211/external/rules_swift_package_manager~0.14.0/swiftpkg/internal/pkginfos.bzl", line 59, column 57, in _get_desc_manifest
		return repository_utils.parsed_json_from_spm_command(
	File "/private/var/tmp/_bazel_krypt/de9844abaa047ea4e0868fa95089f211/external/rules_swift_package_manager~0.14.0/swiftpkg/internal/repository_utils.bzl", line 82, column 49, in _parsed_json_from_spm_command
		json_str = repository_utils.exec_spm_command(
	File "/private/var/tmp/_bazel_krypt/de9844abaa047ea4e0868fa95089f211/external/rules_swift_package_manager~0.14.0/swiftpkg/internal/repository_utils.bzl", line 68, column 13, in _execute_spm_command
		fail(err_msg_tpl.format(
Error in fail: Failed to execute SPM command. name: rules_swift_package_manager~0.14.0~swift_deps~swiftpkg_realm_core, args: ["xcrun", "swift", "package", "describe", "--type", "json"], return_code: 1
error: invalid custom path 'external/catch/src' for target 'Catch2'
.
ERROR: <builtin>: fetching swift_package rule //:rules_swift_package_manager~0.14.0~swift_deps~swiftpkg_realm_core: Traceback (most recent call last):
	File "/private/var/tmp/_bazel_krypt/de9844abaa047ea4e0868fa95089f211/external/rules_swift_package_manager~0.14.0/swiftpkg/internal/swift_package.bzl", line 65, column 28, in _swift_package_impl
		pkg_ctx = pkg_ctxs.read(repository_ctx, directory, env)
	File "/private/var/tmp/_bazel_krypt/de9844abaa047ea4e0868fa95089f211/external/rules_swift_package_manager~0.14.0/swiftpkg/internal/pkg_ctxs.bzl", line 13, column 28, in _read
		pkg_info = pkginfos.get(
	File "/private/var/tmp/_bazel_krypt/de9844abaa047ea4e0868fa95089f211/external/rules_swift_package_manager~0.14.0/swiftpkg/internal/pkginfos.bzl", line 87, column 39, in _get
		desc_manifest = _get_desc_manifest(
	File "/private/var/tmp/_bazel_krypt/de9844abaa047ea4e0868fa95089f211/external/rules_swift_package_manager~0.14.0/swiftpkg/internal/pkginfos.bzl", line 59, column 57, in _get_desc_manifest
		return repository_utils.parsed_json_from_spm_command(
	File "/private/var/tmp/_bazel_krypt/de9844abaa047ea4e0868fa95089f211/external/rules_swift_package_manager~0.14.0/swiftpkg/internal/repository_utils.bzl", line 82, column 49, in _parsed_json_from_spm_command
		json_str = repository_utils.exec_spm_command(
	File "/private/var/tmp/_bazel_krypt/de9844abaa047ea4e0868fa95089f211/external/rules_swift_package_manager~0.14.0/swiftpkg/internal/repository_utils.bzl", line 68, column 13, in _execute_spm_command
		fail(err_msg_tpl.format(
Error in fail: Failed to execute SPM command. name: rules_swift_package_manager~0.14.0~swift_deps~swiftpkg_realm_core, args: ["xcrun", "swift", "package", "describe", "--type", "json"], return_code: 1
error: invalid custom path 'external/catch/src' for target 'Catch2'
.
ERROR: /private/var/tmp/_bazel_krypt/de9844abaa047ea4e0868fa95089f211/external/rules_swift_package_manager~0.14.0~swift_deps~swiftpkg_realm_swift/BUILD.bazel:28:19: @rules_swift_package_manager~0.14.0~swift_deps~swiftpkg_realm_swift//:Realm_modulemap depends on @rules_swift_package_manager~0.14.0~swift_deps~swiftpkg_realm_core//:src_RealmCore_modulemap in repository @rules_swift_package_manager~0.14.0~swift_deps~swiftpkg_realm_core which failed to fetch. no such package '@rules_swift_package_manager~0.14.0~swift_deps~swiftpkg_realm_core//': Failed to execute SPM command. name: rules_swift_package_manager~0.14.0~swift_deps~swiftpkg_realm_core, args: ["xcrun", "swift", "package", "describe", "--type", "json"], return_code: 1
error: invalid custom path 'external/catch/src' for target 'Catch2'
```

**Comment by emartinson** [comment]
Hi! Any updates here?

**Comment by mikhailmulyar** [comment]
`error: invalid custom path 'external/catch/src' for target 'Catch2'`
This issue happens because https://github.com/realm/realm-core has Catch2 dependency as submodule and this path is empty at the moment of resolving. It might be fixed in latest versions of [rules_swift_package_manager](https://github.com/cgrindel/rules_swift_package_manager), but im not sure. I've managed to fix it by implementing patch to _realm-core_ Package file and removing all dependencies of Catch2, because it is required only by test targets.

However after applying patch, there is another issue rising up.
`fail("No header files were provided.")`
It appears in `realm-swift/Realm` and also in `realm-core`.
Also it appears in latest version of [rules_swift_package_manager](https://github.com/cgrindel/rules_swift_package_manager) when trying to add [aws_sdk](https://github.com/aws-amplify/aws-sdk-ios-spm).
It seems in all cases it happens because those sdks create dummy targets in their Package.swift file. Those targets contain only single empty.m file and binary dependencies. So it seems rules_spm can't see any headers while creating module map for those targets.
https://github.com/aws-amplify/aws-sdk-ios-spm/blob/f7e79d076556210b1aec4289cc0028d7ac785959/Package.swift#L193
At the moment i haven't found any option to fix this behaviour.

**Comment by cgrindel** [comment]
Closing as addressable with existing configuration.

The underlying error — `invalid custom path 'external/catch/src' for target 'Catch2'` — is because realm-core uses git submodules (Catch2 lives at `external/catch/src`), and SPM can't resolve that custom path when the submodule isn't checked out.

`swift_package` supports cloning submodules. In bzlmod, configure it per-package:

```starlark
swift_deps.configure_package(
    name = "swiftpkg_realm_core",
    init_submodules = True,
    # recursive_init_submodules = True,  # default
)
```

That should let SPM resolve the Catch2 target and unblock the realm-swift build.

Note on the follow-on @mikhailmulyar mentioned (`No header files were provided` for dummy targets like the aws-sdk-ios-spm binary-dep wrapper targets): that's a separate problem from submodules. If it's still reproducible on a current rules_swift_package_manager release, please open a new issue with a minimal repro so we can track it independently.
